src/devices/Bmxx80/Bmxx80Base.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Device.I2c;
+using System.IO;
+using Iot.Device.Bmxx80.CalibrationData;
+using Iot.Device.Bmxx80.PowerMode;
+using Iot.Device.Bmxx80.Register;
+using Iot.Units;
+
+namespace Iot.Device.Bmxx80
+{
+    /// <summary>
+    /// Represents the core functionality of the Bmxx80 family.
+    /// </summary>
+    public abstract class Bmxx80Base : IDisposable
+    {
+        internal Bmxx80CalibrationData _calibrationData;
+        internal I2cDevice _i2cDevice;
+        internal CommunicationProtocol _communicationProtocol;
+        internal byte _controlRegister;
+
+        internal enum CommunicationProtocol
+        {
+            I2c
+        }
+
+        /// <summary>
+        /// The variable _temperatureFine carries a fine resolution temperature value over to the
+        /// pressure compensation formula and could be implemented as a global variable.
+        /// </summary>
+        protected int TemperatureFine;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Bmxx80Base"/> class.
+        /// </summary>
+        /// <param name="deviceId">The ID of the device.</param>
+        /// <param name="i2cDevice">The <see cref="I2cDevice"/> to create with.</param>
+        /// <exception cref="ArgumentNullException"></exception>
+        /// <exception cref="IOException"></exception>
+        protected Bmxx80Base(byte deviceId, I2cDevice i2cDevice)
+        {
+            _i2cDevice = i2cDevice ?? throw new ArgumentNullException(nameof(i2cDevice));
+            _i2cDevice.WriteByte((byte)Bmxx80Register.CHIPID);
+
+            byte readSignature = _i2cDevice.ReadByte();
+
+            if (readSignature != deviceId)
+            {
+                throw new IOException($"Unable to find a chip with id {deviceId}");
+            }
+        }
+
+        /// <summary>
+        /// Sets the pressure sampling.
+        /// </summary>
+        /// <param name="sampling">The <see cref="Sampling"/> to set.</param>
+        public void SetPressureSampling(Sampling sampling)
+        {
+            byte status = Read8BitsFromRegister(_controlRegister);
+            status = (byte)(status & 0b1110_0011);
```

Comment:
while I'm fine with this, ususally we do something like:
```c#
status &= (byte)~(0b111 << 2);
```

Reply:
This looks like it would need to be `status &= unchecked((byte)~(0b111 << 2));`, or am I missing something obvious? 😅

Reply:
I wrestled with the same issue and that's down to how C# handles byte values, so your ```0b111``` gets treated as an integer and the unchecked bitwise complement will therefore also return an integer __which is negative__ so this will give you a wrong result.

The way you can remedy this is to define the ```0b111``` explicitly as a byte or what I did was to define Masks in a enum as byte values which also removes a bit of redundancy when you are using those values for both getters and setters.

This helped me wrap my head around this: https://www.oreilly.com/library/view/c-cookbook/0596003390/ch01s04.html

Reply:
To be a little more precise it's not exactly down to how C# handles byte values but that 0b111 is just another way of expressing an int value.

Reply:
@georgemathieson: `status &= unchecked((byte)~(0b111 << 2));` looks correct, although I think the original version is more readable in this case so I'd leave as is

Reply:
oh @krwq is right, ignore my comment I had a wrong understanding of the unchecked syntax 😨. Just tested it and your code would give the right result, my bad!

Reply:
Shall leave as is for now as I agree with the original being more readable.